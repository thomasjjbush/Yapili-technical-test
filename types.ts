import { ReactNode } from 'react';
import { AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import HtmlWebPackPlugin from 'html-webpack-plugin';
import { Json } from 'enzyme-to-json';
import { MockStoreEnhanced } from 'redux-mock-store';

export type Action = ThunkAction<Promise<AnyAction | void>, Store, unknown, AnyAction>;

export interface AspectRatioProps {
    children: ReactNode;
    ratio: AspectRatios;
}

export enum AspectRatios {
    '1:1' = 100,
    '16:9' = 56.25,
    '4:3' = 75,
}

export interface Character {
    comics?: CharecterCollection;
    description?: string;
    error?: boolean;
    events?: CharecterCollection;
    id?: number;
    modified?: string;
    name?: string;
    resourceURI?: string;
    series?: CharecterCollection;
    stories?: CharecterCollection;
    thumbnail?: {
        extension: string;
        path: string;
    };
}

export enum CharacterActions {
    CLEARED = 'character/CLEARED',
    ERROR = 'character/ERROR',
    LOADED = 'character/LOADED',
}

interface CharecterCollection {
    available?: number;
    collectionURI?: string;
    items: Resource[];
    returned?: number;
}

export enum Dimensions {
    LANDSCAPE = 'landscape',
    STANDARD = 'standard',
}

export type Dispatch = ThunkDispatch<Store, void, AnyAction>;

export interface FallbackImageProps {
    src: string;
    width?: string;
    height?: string;
}

export interface Home extends Partial<MarvelResponse> {
    error?: boolean;
    results?: Character[];
}

export enum HomeActions {
    ERROR = 'home/ERROR',
    LOADED = 'home/LOADED',
    LOADING = 'home/LOADING',
}

export interface ImageProps {
    dimension?: Dimensions;
    extension: string;
    height?: string;
    path: string;
    width?: string;
}

export interface MarvelResponse {
    count: number;
    limit: number;
    offset: number;
    results: any[];
    total: number;
}

export type MockStore = MockStoreEnhanced<Partial<Store>, Dispatch>;

export interface Options {
    [key: string]: string | number;
    apikey?: string;
    limit?: number;
    offset?: number;
}

export interface PaginatedButtonsProps {
    currentPage: number;
    numberOfButtons: number;
    numberOfPages: number;
    skip?: number;
    to: (i: number) => Partial<Location>;
}

export interface Resource {
    resourceURI: string;
    name: string;
    type?: string;
}

export interface SerializerMap extends Json {
    node: {
        type: {
            componentStyle?: {
                baseStyle?: {
                    rules: (string | ((props: StyledProps) => string))[];
                };
                rules: (string | ((props: StyledProps) => string))[];
            };
        };
    };
}

export interface Store {
    character: Character;
    home: Home;
}

export type StyledProps<Props = {}> = { theme: Theme } & Props;

export interface Theme {
    breakpoints: {
        desktop: number;
        tabletLandscape: number;
        tabletPortrait: number;
    };
    colors: {
        darkGrey: string;
        marvelRed: string;
        mediumGrey: string;
    };
}

export interface WebpackConfig {
    devServer: { historyApiFallback: boolean; open: boolean };
    entry: string[];
    mode: string;
    module: {
        rules: WebpackRule[];
    };
    output: { path?: string; publicPath: string };
    plugins: HtmlWebPackPlugin[];
    resolve: { extensions: string[] };
}

interface WebpackRule {
    exclude?: (RegExp | string)[];
    test: RegExp;
    use: {
        loader: string;
        options?: { plugins: string[] };
    };
}
