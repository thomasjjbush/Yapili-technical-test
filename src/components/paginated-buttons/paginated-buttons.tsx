import React, { FC, ReactElement } from 'react';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { PaginatedButtonsProps as Props, StyledProps } from './../../../types';

const Button = styled(Link)<StyledProps<{ $active?: boolean; disabled?: boolean }>>`
    background-color: ${({ $active, theme }): string => ($active ? theme.colors.marvelRed : 'transparent')};
    color: white;
    padding: 10px;
    text-decoration: none;

    &[disabled] {
        cursor: not-allowed;
        opacity: 0.5;
    }
`;

const Container = styled.div<StyledProps>`
    background-color: ${({ theme }) => theme.colors.darkGrey};
    border: solid 2px ${({ theme }) => theme.colors.marvelRed};
    border-radius: 10px;
    display: inline-flex;
    margin-top: 30px;
    overflow: hidden;
`;

export const PaginatedButtons: FC<Props> = ({
    currentPage,
    numberOfButtons,
    numberOfPages,
    skip,
    to,
}: Props): ReactElement => {
    let firstButton = Math.max(currentPage - Math.floor(numberOfButtons / 2), 1);

    if (numberOfPages - firstButton < numberOfButtons - 1) {
        firstButton = firstButton - (numberOfButtons - 1 - (numberOfPages - firstButton));
    }

    return (
        <Container>
            {skip && (
                <Button data-test-id="skip" disabled={currentPage - skip < 1} to={to(currentPage - skip)}>
                    -{skip}
                </Button>
            )}
            {[...new Array(Math.min(numberOfButtons, numberOfPages))].map((_, i) => {
                const newPage = firstButton + i;
                return (
                    <Button data-test-id="button" $active={newPage === currentPage} key={newPage} to={to(newPage)}>
                        {newPage}
                    </Button>
                );
            })}
            {skip && (
                <Button data-test-id="skip" disabled={currentPage + skip > numberOfPages} to={to(currentPage + skip)}>
                    +{skip}
                </Button>
            )}
        </Container>
    );
};
