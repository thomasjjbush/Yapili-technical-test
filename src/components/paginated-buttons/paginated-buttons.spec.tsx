import React from 'react';
import { shallow } from 'enzyme';
import { PaginatedButtons } from './paginated-buttons';
import { PaginatedButtonsProps as Props } from '../../../types';

describe('PaginatedButtons', () => {
    let props: Props;

    beforeEach(() => {
        props = {
            currentPage: 1,
            numberOfButtons: 5,
            numberOfPages: 10,
            to: jest.fn((i: number) => ({ search: `something=${i}` })),
        };
    });

    it('should match snapshot', () => {
        const wrapper = shallow(<PaginatedButtons {...props} skip={5} />);
        expect(wrapper).toMatchSnapshot();
    });

    it.each`
        currentPage | expectedButtons
        ${1}        | ${['1', '2', '3', '4', '5']}
        ${2}        | ${['1', '2', '3', '4', '5']}
        ${3}        | ${['1', '2', '3', '4', '5']}
        ${4}        | ${['2', '3', '4', '5', '6']}
        ${5}        | ${['3', '4', '5', '6', '7']}
        ${6}        | ${['4', '5', '6', '7', '8']}
        ${7}        | ${['5', '6', '7', '8', '9']}
        ${8}        | ${['6', '7', '8', '9', '10']}
        ${9}        | ${['6', '7', '8', '9', '10']}
        ${10}       | ${['6', '7', '8', '9', '10']}
    `('should render correct array of buttons for current page of $currentPage', ({ currentPage, expectedButtons }) => {
        const wrapper = shallow(<PaginatedButtons {...props} currentPage={currentPage} />);
        const buttons = wrapper.find({ 'data-test-id': 'button' });

        buttons.forEach((button, i) => {
            expect(button.text()).toBe(expectedButtons[i]);
            expect(button.prop('$active')).toBe(Number(button.text()) === currentPage);
        });
    });

    it('should disabled skip forward button if skip value is greater than remaining buttons', () => {
        const wrapper = shallow(<PaginatedButtons {...props} currentPage={6} skip={5} />);
        expect(wrapper.find({ 'data-test-id': 'skip' }).last().prop('disabled')).toBe(true);
    });

    it('should render less buttons if there are less pages than requested buttons to display', () => {
        const wrapper = shallow(<PaginatedButtons {...props} numberOfPages={4} />);
        expect(wrapper.find({ 'data-test-id': 'button' }).length).toBe(4);
    });
});
