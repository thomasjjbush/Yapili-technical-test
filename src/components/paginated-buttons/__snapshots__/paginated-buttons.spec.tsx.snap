// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginatedButtons should match snapshot 1`] = `
<styled.div
  styles="
    background-color: theme.colors.darkGrey;
    border: solid 2px theme.colors.marvelRed;
    border-radius: 10px;
    display: inline-flex;
    margin-top: 30px;
    overflow: hidden;
"
>
  <Styled(Link)
    data-test-id="skip"
    disabled={true}
    styles="
    background-color: transparent;
    color: white;
    padding: 10px;
    text-decoration: none;

    &[disabled] {
        cursor: not-allowed;
        opacity: 0.5;
    }
"
    to={
      Object {
        "search": "something=-4",
      }
    }
  >
    -
    5
  </Styled(Link)>
  <Styled(Link)
    $active={true}
    data-test-id="button"
    key="1"
    styles="
    background-color: theme.colors.marvelRed;
    color: white;
    padding: 10px;
    text-decoration: none;

    &[disabled] {
        cursor: not-allowed;
        opacity: 0.5;
    }
"
    to={
      Object {
        "search": "something=1",
      }
    }
  >
    1
  </Styled(Link)>
  <Styled(Link)
    $active={false}
    data-test-id="button"
    key="2"
    styles="
    background-color: transparent;
    color: white;
    padding: 10px;
    text-decoration: none;

    &[disabled] {
        cursor: not-allowed;
        opacity: 0.5;
    }
"
    to={
      Object {
        "search": "something=2",
      }
    }
  >
    2
  </Styled(Link)>
  <Styled(Link)
    $active={false}
    data-test-id="button"
    key="3"
    styles="
    background-color: transparent;
    color: white;
    padding: 10px;
    text-decoration: none;

    &[disabled] {
        cursor: not-allowed;
        opacity: 0.5;
    }
"
    to={
      Object {
        "search": "something=3",
      }
    }
  >
    3
  </Styled(Link)>
  <Styled(Link)
    $active={false}
    data-test-id="button"
    key="4"
    styles="
    background-color: transparent;
    color: white;
    padding: 10px;
    text-decoration: none;

    &[disabled] {
        cursor: not-allowed;
        opacity: 0.5;
    }
"
    to={
      Object {
        "search": "something=4",
      }
    }
  >
    4
  </Styled(Link)>
  <Styled(Link)
    $active={false}
    data-test-id="button"
    key="5"
    styles="
    background-color: transparent;
    color: white;
    padding: 10px;
    text-decoration: none;

    &[disabled] {
        cursor: not-allowed;
        opacity: 0.5;
    }
"
    to={
      Object {
        "search": "something=5",
      }
    }
  >
    5
  </Styled(Link)>
  <Styled(Link)
    data-test-id="skip"
    disabled={false}
    styles="
    background-color: transparent;
    color: white;
    padding: 10px;
    text-decoration: none;

    &[disabled] {
        cursor: not-allowed;
        opacity: 0.5;
    }
"
    to={
      Object {
        "search": "something=6",
      }
    }
  >
    +
    5
  </Styled(Link)>
</styled.div>
`;
